pilot/pkg/controller/workloadentry/workloadentry_controller.go

```diff
@@ -332,6 +374,22 @@ func (c *Controller) shouldCleanupEntry(wle config.Config) bool {
 		return false
 	}
 
+	// If there is ConnectedAtAnnotation set, don't cleanup this workload entry.
+	// This may happen when the workload fast reconnects to the same istiod.
+	// 1. disconnect: the workload entry has been updated
+	// 2. connect: but the patch is based on the old workloadentry because of the propagation latency.
+	// So in this case the `DisconnectedAtAnnotation` is still there and the cleanup procedure will go on.
+	connTime := wle.Annotations[ConnectedAtAnnotation]
```

Comment:
This approach would probably work. I think we would also need the merge patch, or to add retry on `errors.IsInvalid`. If we delete the annotation on line 290 _after_ we Get the entry for the patch, the patch will be confused whether to use op: add or op: replace.

Reply:
Without merge-patch, the json-patch can return error, and then the stream will be recreated. We depend on the clientside retry. 

If merge patch applied, it may mask potential races？ I am not sure how much effort would be needed to handle that. This is the main concern i have.

Reply:
If we do retries, we probably want to leave as many annotations set as possible so that we can determine ordering. With the retry approach, we have a true "last-write-wins" on the entire object. Maybe this is easier. 

Using merge means we avoid retries which seems easier to think about when we have this concurrent access. That approach is optimistic for connections, but discards disconnections more easily. There are a few more things to make it work included in my PR.

If we're going with the retry approach, I don't see a reason to use patch at all. If anything that would mask races, since it _may or may not_ succeed even with a concurrent modification, even without merge patch. It's probably best to change this to Update for the purposes of this PR.

Reply:
This pr is not to care about whether patch, merge patch or update. The issue it tries to resolve can be not be done by purely change `patch, merge patch or update`. 

We can discuss that in a new issue.

Reply:
If the rest approach relies on setting/unsetting annotations, you're likely to completely fail out on the `Patch` when not using merge patch. I'm fine with leaving the existing patch here, but adding retries. 

If we're relying on retries anyway,  it seems easier to just use update + [k8s.io/client-go/util/retry](https://godoc.org/k8s.io/client-go/util/retry). That way, we retry in all conflicts instead of some subset, making things easier to reason about and debug.

Reply:
yeah, `update` sgtm.